File: rust/src/discord.rs
use reqwest::Client;
use serde::Serialize;
use serde_json::Value;
use std::sync::RwLock;
use std::time::Duration;
use tokio::time::sleep;

pub const DISCORD_API_BASE: &str = "https://discord.com/api/v10";
const MAX_RETRIES: u32 = 3;

static DISCORD_API_BASE_OVERRIDE: RwLock<Option<String>> = RwLock::new(None);

fn api_base() -> String {
    if let Ok(guard) = DISCORD_API_BASE_OVERRIDE.read() {
        if let Some(base) = guard.as_ref() {
            return base.clone();
        }
    }
    DISCORD_API_BASE.to_string()
}

/// Override Discord API base URL for testing. Pass None to reset.
pub fn _set_discord_api_base(base: Option<String>) {
    *DISCORD_API_BASE_OVERRIDE.write().unwrap() = base;
}

#[derive(Serialize)]
struct SendMessagePayload {
    content: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    embeds: Option<Vec<Value>>,
}

pub async fn send_message(
    token: &str,
    channel_id: &str,
    text: &str,
    client: &Client,
) -> Result<Value, String> {
    let payload = SendMessagePayload {
        content: text.to_string(),
        embeds: None,
    };
    post_message(token, channel_id, &payload, client).await
}

pub async fn send_card(
    token: &str,
    channel_id: &str,
    text: &str,
    card: &Value,
    client: &Client,
) -> Result<Value, String> {
    let embeds = card
        .as_array()
        .cloned()
        .unwrap_or_else(|| vec![card.clone()]);

    let payload = SendMessagePayload {
        content: text.to_string(),
        embeds: Some(embeds),
    };
    post_message(token, channel_id, &payload, client).await
}

async fn post_message(
    token: &str,
    channel_id: &str,
    payload: &SendMessagePayload,
    client: &Client,
) -> Result<Value, String> {
    let url = format!("{}/channels/{}/messages", api_base(), channel_id);
    let payload_json = serde_json::to_string(payload)
        .map_err(|e| format!("serialize payload: {}", e))?;

    let mut attempt = 0;
    loop {
        let resp = client
            .post(&url)
            .header("Authorization", format!("Bot {}", token))
            .header("Content-Type", "application/json")
            .body(payload_json.clone())
            .send()
            .await
            .map_err(|e| format!("send message: {}", e))?;

        let status = resp.status();

        if status.as_u16() == 429 && attempt < MAX_RETRIES {
            let retry_after = resp
                .headers()
                .get("Retry-After")
                .and_then(|h| h.to_str().ok())
                .and_then(|s| s.parse::<u64>().ok())
                .unwrap_or(1);

            sleep(Duration::from_secs(retry_after)).await;
            attempt += 1;
            continue;
        }

        let body = resp.text().await.map_err(|e| format!("read response: {}", e))?;

        if !status.is_success() {
            return Err(format!("HTTP {}: {}", status.as_u16(), body));
        }
        return serde_json::from_str(&body).map_err(|e| format!("parse response: {}", e));
    }
}

File: python/src/wgrok/discord.py
"""Thin Discord REST client for sending messages and embeds."""

from __future__ import annotations

import asyncio

import aiohttp

DISCORD_API_BASE = "https://discord.com/api/v10"
MAX_RETRIES = 3


def _headers(token: str) -> dict:
    return {"Authorization": f"Bot {token}", "Content-Type": "application/json"}


def _messages_url(channel_id: str) -> str:
    return f"{DISCORD_API_BASE}/channels/{channel_id}/messages"


async def _manage_session(session, func):
    """Run func with session, closing it afterward if we created it."""
    owns_session = session is None
    if owns_session:
        session = aiohttp.ClientSession()
    try:
        return await func(session)
    finally:
        if owns_session:
            await session.close()


async def _request_with_retry(session, method, url, headers, **kwargs):
    """Execute an HTTP request with Retry-After handling for 429 responses."""
    for attempt in range(MAX_RETRIES + 1):
        async with session.request(method, url, headers=headers, **kwargs) as resp:
            if resp.status == 429:
                if attempt >= MAX_RETRIES:
                    resp.raise_for_status()
                try:
                    retry_after = min(int(resp.headers.get("Retry-After", "1")), 300)
                except (ValueError, TypeError):
                    retry_after = 1
                await asyncio.sleep(retry_after)
                continue
            resp.raise_for_status()
            return await resp.json()
    msg = f"HTTP request to {url} failed after {MAX_RETRIES + 1} attempts"
    raise aiohttp.ClientResponseError(None, None, message=msg, status=429)


async def send_message(
    token: str,
    channel_id: str,
    text: str,
    session: aiohttp.ClientSession | None = None,
) -> dict:
    """Send a text-only Discord message to a channel."""
    url = _messages_url(channel_id)
    payload = {"content": text}

    async def _do(s):
        return await _request_with_retry(s, "POST", url, _headers(token), json=payload)

    return await _manage_session(session, _do)


async def send_card(
    token: str,
    channel_id: str,
    text: str,
    card: dict,
    session: aiohttp.ClientSession | None = None,
) -> dict:
    """Send a Discord message with an embed.

    Args:
        text: Message content text.
        card: Discord embed object (passed as first element of 'embeds' array).
    """
    embeds = card if isinstance(card, list) else [card]
    url = _messages_url(channel_id)
    payload = {"content": text, "embeds": embeds}

    async def _do(s):
        return await _request_with_retry(s, "POST", url, _headers(token), json=payload)

    return await _manage_session(session, _do)
